src/ghci/mod.rs

```diff
@@ -487,38 +482,54 @@ impl Ghci {
     /// Restart the `ghci` session.
     #[instrument(skip_all, level = "debug")]
     async fn restart(&mut self) -> miette::Result<()> {
-        self.run_hooks(LifecycleEvent::Restart(hooks::When::Before))
+        let mut log = CompilationLog::default();
+
+        self.run_hooks(LifecycleEvent::Restart(hooks::When::Before), &mut log)
             .await?;
         self.stop().await?;
         let new = Self::new(self.shutdown.clone(), self.opts.clone()).await?;
         let _ = std::mem::replace(self, new);
-        self.initialize().await?;
-        self.run_hooks(LifecycleEvent::Restart(hooks::When::After))
+        self.initialize(&mut log).await?;
+        self.run_hooks(LifecycleEvent::Restart(hooks::When::After), &mut log)
             .await?;
+
+        self.write_error_log(&log).await?;
         Ok(())
     }
 
     /// Run the user provided test command.
     #[instrument(skip_all, level = "debug")]
-    async fn test(&mut self) -> miette::Result<()> {
-        self.stdin.set_mode(&mut self.stdout, Mode::Testing).await?;
-        self.run_hooks(LifecycleEvent::Test).await?;
+    async fn test(&mut self, log: &mut CompilationLog) -> miette::Result<()> {
+        self.run_hooks(LifecycleEvent::Test, log).await?;
         Ok(())
     }
 
     /// Run the eval commands, if enabled.
     #[instrument(skip_all, level = "debug")]
-    async fn eval(&mut self) -> miette::Result<()> {
+    async fn eval(&mut self, log: &mut CompilationLog) -> miette::Result<()> {
         if !self.opts.enable_eval {
             return Ok(());
         }
 
-        for (path, commands) in &self.eval_commands {
+        // TODO: This `clone` is ugly but I can't get the borrow checker to accept it otherwise.
+        // Might be more efficient to swap it out for a default, but then it gets trickier to
+        // restore the old value when the function returns.
+        for (path, commands) in self.eval_commands.clone() {
             for command in commands {
                 tracing::info!("{path}:{command}");
-                let module_name = self.search_paths.path_to_module(path)?;
+                // If the `module` was already compiled, `ghci` may have loaded the interface file instead
+                // of the interpreted bytecode, giving us this error message when we attempt to
+                // load the top-level scope with `:module + *{module}`:
+                //
+                //     module 'Mercury.Typescript.Golden' is not interpreted
+                //
+                // We use `:add *{module}` to force interpreting the module. We do this here instead of in
+                // `add_module` to save time if eval commands aren't used (or aren't needed for a
+                // particular module).
+                self.interpret_module(&path, log).await?;
```

Comment:
Here's the actual bugfix, making sure the modules are interpreted and using the module path (instead of the module name) to import them.